x/paychan/handler.go

```diff
@@ -0,0 +1,296 @@
+	if _, err := h.validate(ctx, db, tx); err != nil {
+		return res, err
+	}
+
+	res.GasAllocated += createPaymentChannelCost
+	return res, nil
+}
+
+func (h *createPaymentChannelHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*CreatePaymentChannelMsg, error) {
+	rmsg, err := tx.GetMsg()
+	if err != nil {
+		return nil, err
+	}
+	msg, ok := rmsg.(*CreatePaymentChannelMsg)
+	if !ok {
+		return nil, errors.ErrUnknownTxType(rmsg)
+	}
+
+	if err := msg.Validate(); err != nil {
+		return msg, err
+	}
+
+	// Ensure that the timeout is in the future.
+	if height, _ := weave.GetHeight(ctx); msg.Timeout <= height {
+		return msg, ErrInvalidTimeout(msg.Timeout)
+	}
+
+	if !h.auth.HasAddress(ctx, msg.SenderPublicKey.Address()) {
+		return msg, errors.ErrUnauthorized()
+	}
+
+	return msg, nil
+}
+
+func (h *createPaymentChannelHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.DeliverResult, error) {
+	var res weave.DeliverResult
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return res, err
+	}
+
+	obj, err := h.bucket.Create(db, &PaymentChannel{
+		SenderPublicKey: msg.SenderPublicKey,
+		Recipient:       msg.Recipient,
+		Total:           msg.Total,
+		Timeout:         msg.Timeout,
+		Memo:            msg.Memo,
+		Transferred: &x.Coin{
+			Whole:      0,
+			Fractional: 0,
+			Ticker:     msg.Total.Ticker,
+		},
+	})
+	if err != nil {
+		return res, err
+	}
+
+	// Move coins from sender account and deposit total amount available on
+	// that channels account.
+	src := msg.SenderPublicKey.Address()
+	dst := paymentChannelAccount(obj.Key())
+	if err := h.cash.MoveCoins(db, src, dst, *msg.Total); err != nil {
+		return res, err
+	}
+
+	res.Data = obj.Key()
+	return res, err
+}
+
+type transferPaymentChannelHandler struct {
+	auth   x.Authenticator
+	bucket PaymentChannelBucket
+	cash   cash.Controller
+}
+
+var _ weave.Handler = (*transferPaymentChannelHandler)(nil)
+
+func (h *transferPaymentChannelHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.CheckResult, error) {
+	var res weave.CheckResult
+	if _, err := h.validate(ctx, db, tx); err != nil {
+		return res, err
+	}
+	res.GasAllocated += transferPaymentChannelCost
+	return res, nil
+}
+
+func (h *transferPaymentChannelHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*TransferPaymentChannelMsg, error) {
+	rmsg, err := tx.GetMsg()
+	if err != nil {
+		return nil, err
+	}
+	msg, ok := rmsg.(*TransferPaymentChannelMsg)
+	if !ok {
+		return nil, errors.ErrUnknownTxType(rmsg)
+	}
+
+	if err := msg.Validate(); err != nil {
+		return msg, err
+	}
+
+	// TODO(husio) validate that the payment is done for the right chain
+
+	pc, err := h.bucket.GetPaymentChannel(db, msg.Payment.ChannelId)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check signature to ensure the message was not altered.
+	raw, err := msg.Payment.Marshal()
+	if err != nil {
+		return nil, fmt.Errorf("serialize payment: %s", err)
```

Comment:
req: please use `errors.Wrap(err, "serialize payment")` instead. See https://github.com/pkg/errors#adding-context-to-an-error